File: scripts/reorder_lessons.py
import json
import os
import requests

def reorder_lesson_words():
    base_dir = "/Users/prityushjhaveri/Projects/Year 4/cantonese"
    output_file = os.path.join(base_dir, "cantonese-app", "data", "chapters.json")
    
    with open(output_file, 'r', encoding='utf-8') as f:
        chapters_data = json.load(f)

    # Group by lesson
    lesson_groups = {}
    for word in chapters_data:
        lesson = word['lessonStr']
        if lesson not in lesson_groups:
            lesson_groups[lesson] = []
        lesson_groups[lesson].append(word)

    print(f"Loaded {len(chapters_data)} words across {len(lesson_groups)} lessons.")

    # We will use Poe API to logically group words within each lesson
    # Only if the lesson has more than 3 words to make it worth it
    
    api_key = "C0CZJ6kdNd9fRWaJP1wiMefMGd_Zsge38axkhB1oJvg"
    headers = {
        'Authorization': f'Bearer {api_key}',
        'Content-Type': 'application/json'
    }

    reordered_data = []

    for lesson, words in lesson_groups.items():
        if len(words) <= 3:
            reordered_data.extend(words)
            continue
            
        print(f"Reordering {lesson} ({len(words)} words)...")
        
        # Prepare lightweight list for LLM to reorder
        word_list = [{"id": w['number'], "eng": w['english'], "pos": w['pos']} for w in words]
        
        prompt = f"""
You are an expert curriculum designer. Please reorder the following list of vocabulary words so that words of similar categories, themes, or grammatical functions appear sequentially together (e.g. all countries together, all food items together, all pronouns together). 

Here is the list:
{json.dumps(word_list, ensure_ascii=False, indent=2)}

Please return ONLY a JSON array containing the exactly identical objects, but in the new, logically grouped order. Do not change the keys or values. Do not use markdown backticks, return raw JSON array only.
"""
        try:
            response = requests.post('https://api.poe.com/v1/chat/completions', headers=headers, json={
                "model": "Gemini-3-Flash",
                "messages": [
                    {"role": "user", "content": prompt}
                ],
                "temperature": 0.1
            })
            
            if response.ok:
                resp_json = response.json()
                content = resp_json['choices'][0]['message']['content'].strip()
                
                if content.startswith('```json'):
                    content = content[7:-3].strip()
                elif content.startswith('```'):
                    content = content[3:-3].strip()
                    
                new_order = json.loads(content)
                
                # Reconstruct full word objects based on the new ID order
                ordered_full_words = []
                for basic_word in new_order:
                    # Find matching word in original list
                    match = next((w for w in words if w['number'] == basic_word['id']), None)
                    if match:
                        ordered_full_words.append(match)
                
                # Handle any that were missed by the LLM
                for w in words:
                    if w not in ordered_full_words:
                        ordered_full_words.append(w)
                        
                reordered_data.extend(ordered_full_words)
            else:
                print(f"Failed to reorder {lesson}, falling back to original. Status: {response.status_code}")
                reordered_data.extend(words)
        except Exception as e:
            print(f"Error reordering {lesson}: {e}")
            reordered_data.extend(words)

    # Re-apply sequential numbering based on the new order
    for idx, word in enumerate(reordered_data):
        word['number'] = idx + 1

    # Save it back
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(reordered_data, f, ensure_ascii=False, indent=2)

    print(f"Successfully reordered and saved {len(reordered_data)} words.")

if __name__ == "__main__":
    reorder_lesson_words()

File: scripts/reorder_lessons.js
const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');

async function reorderLessonWords() {
    const baseDir = "/Users/prityushjhaveri/Projects/Year 4/cantonese";
    const outputFile = path.join(baseDir, "cantonese-app", "data", "chapters.json");

    if (!fs.existsSync(outputFile)) {
        console.error("chapters.json not found!");
        return;
    }

    const rawData = fs.readFileSync(outputFile, 'utf8');
    let chaptersData = JSON.parse(rawData);

    // Group by lesson
    const lessonGroups = {};
    for (const word of chaptersData) {
        const lesson = word.lessonStr;
        if (!lessonGroups[lesson]) {
            lessonGroups[lesson] = [];
        }
        lessonGroups[lesson].push(word);
    }

    console.log(`Loaded ${chaptersData.length} words across ${Object.keys(lessonGroups).length} lessons.`);

    const apiKey = "C0CZJ6kdNd9fRWaJP1wiMefMGd_Zsge38axkhB1oJvg";
    const headers = {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json'
    };

    let reorderedData = [];

    for (const [lesson, words] of Object.entries(lessonGroups)) {
        if (words.length <= 3) {
            reorderedData.push(...words);
            continue;
        }

        console.log(`Reordering ${lesson} (${words.length} words)...`);

        const wordList = words.map(w => ({ word: w.english, pos: w.pos }));
        
        const prompt = `
You are an expert curriculum designer. Please reorder the following list of vocabulary words so that words of similar categories, themes, grammatical functions, or logical pairs appear sequentially together (e.g. all countries together, all numbers together, all time words together, question words together). 

Here is the list:
${JSON.stringify(wordList, null, 2)}

Return ONLY a flat JSON array of strings containing the exact 'word' (English) values, but in the new, logically grouped order. Do not change the strings, do not list the POS, just an array of the English strings. Do not use markdown backticks, return raw flat JSON array only.
`;

        try {
            const response = await fetch('https://api.poe.com/v1/chat/completions', {
                method: 'POST',
                headers: headers,
                body: JSON.stringify({
                    model: "Gemini-3-Flash",
                    messages: [{ role: "user", content: prompt }],
                    temperature: 0.1
                })
            });

            if (response.ok) {
                const respJson = await response.json();
                let content = respJson.choices[0].message.content.trim();

                if (content.startsWith('```json')) {
                    content = content.substring(7, content.length - 3).trim();
                } else if (content.startsWith('```')) {
                    content = content.substring(3, content.length - 3).trim();
                }

                let newOrderStrings;
                try {
                     newOrderStrings = JSON.parse(content);
                } catch(e) {
                     console.error("Failed to parse LLM array. Falling back.");
                     reorderedData.push(...words);
                     continue;
                }

                const orderedFullWords = [];
                const usedWords = new Set();
                
                // Construct new objects based on the returned English string
                for (const englishWord of newOrderStrings) {
                    const match = words.find(w => w.english === englishWord && !usedWords.has(w));
                    if (match) {
                        orderedFullWords.push(match);
                        usedWords.add(match);
                    }
                }

                // Handle any words that the LLM forgot or hallucinated out
                for (const w of words) {
                    if (!usedWords.has(w)) {
                        orderedFullWords.push(w);
                    }
                }

                reorderedData.push(...orderedFullWords);
            } else {
                console.log(`Failed to reorder ${lesson}, falling back to original. Status: ${response.status}`);
                reorderedData.push(...words);
            }
        } catch (e) {
            console.error(`Error reordering ${lesson}:`, e.message);
            reorderedData.push(...words);
        }
    }

    // Re-apply sequential numbering based on the new order
    reorderedData = reorderedData.map((word, index) => {
        return { ...word, number: index + 1 };
    });

    // Save it back
    fs.writeFileSync(outputFile, JSON.stringify(reorderedData, null, 2), 'utf8');

    console.log(`Successfully reordered and saved ${reorderedData.length} words.`);
}

reorderLessonWords();
